feat: add strategy registry

steffenix opened the pull request.

fubuloubu commented on contracts/StrategyRegistry.vy:
```diff
@@ -0,0 +1,70 @@
+    numReleases: uint256 
+    versions: address[20]
+
+strategiesReleases: public(HashMap[bytes32, StrategyInfo])
```
`String[64]` can be a key. Saves a hash

fubuloubu replied:
Also, not really sure that using the strategy's name is all that safe, but I guess depends on who can add it and what the outcomes are if it's incorrect.

poolpitako replied:
I think it's fine, it can be based on github repo. For example:
`yearnV2-generic-lender-strat` for https://github.com/Grandthrax/yearnV2-generic-lender-strat
`SingleSidedCurve` for https://github.com/Grandthrax/SingleSidedCurve

I guess we can also normalize repo's names

steffenix replied:
What about different versions of the vault, should the name includes the vault version?

fubuloubu replied:
If we feel strongly that a string is the best domain separator to use here, then yeah this might make sense to do. The norms around "how" the string is set can be decided off-chain, and potentially evolve over time (with some discipline, of course)

fubuloubu commented on contracts/StrategyRegistry.vy:
```diff
@@ -0,0 +1,70 @@
+interface Strategy:
+    def name() -> String[64]: view
+
+struct StrategyInfo:
+    name: String[64]
+    numReleases: uint256 
+    versions: address[20]
+
+strategiesReleases: public(HashMap[bytes32, StrategyInfo])
```
```suggestion
# Strategy.name => release_id
strategiesNumReleases: public(HashMap[String[64], uint256])
# Strategy.name => release_id => Strategy.address
strategyVersions: public(HashMap[String[64], HashMap[uint256, address]])
```

fubuloubu commented on contracts/StrategyRegistry.vy:
```diff
@@ -0,0 +1,70 @@
+    log NewGovernance(msg.sender)
+
+@external
+def addNewRelease(strategy :address, name: String[64] = ""):
+    """
+    @notice add a new version of a strategy
+    """
+    assert msg.sender == self.pendingGovernance  # dev: unauthorized
+    strategyName: String[64] = name 
+    if name == "":
+        strategyName = Strategy(strategy).name()
+    strategyHash: bytes32 = sha256(strategyName)
+    if self.strategiesReleases[strategyHash].numReleases == 0:
+	    self.strategiesReleases[strategyHash] = StrategyInfo({
+		    name: strategyName,
+		    numReleases: 0,
+		    versions: [ZERO_ADDRESS, ZERO_ADDRESS, ZERO_ADDRESS, ZERO_ADDRESS, ZERO_ADDRESS, ZERO_ADDRESS, ZERO_ADDRESS, ZERO_ADDRESS, ZERO_ADDRESS, ZERO_ADDRESS, ZERO_ADDRESS, ZERO_ADDRESS, ZERO_ADDRESS, ZERO_ADDRESS, ZERO_ADDRESS, ZERO_ADDRESS, ZERO_ADDRESS, ZERO_ADDRESS, ZERO_ADDRESS, ZERO_ADDRESS]
+	    })
+    numReleases: uint256 = self.strategiesReleases[strategyHash].numReleases
+    self.strategiesReleases[strategyHash].numReleases = numReleases + 1
+    self.strategiesReleases[strategyHash].versions[numReleases] = strategy
```
what's the usecase for tracking multiple releases of the same strategy name?

steffenix replied:
So I tough that should work like the registry and keep multiple versions for the same name.
What we need is to handle multiple vault versions for the same strategy name, this can be handled by adding the vault version to the name of the strategy or we save it as part of the hashmap

fubuloubu replied:
Don't forget that some of this information might already exist in the Vault registry

steffenix replied:
Can't do that because Dictionary keys must be a base type.

fubuloubu replied:
ah well, `base32` should work.

perhaps should add the ability for any type that can be hashed to be used as a key, since that's really the requirement here.

poolpitako commented on contracts/StrategyRegistry.vy:
```diff
@@ -0,0 +1,70 @@
+    """
+    @notice add a new version of a strategy
+    """
+    assert msg.sender == self.pendingGovernance  # dev: unauthorized
```
why `pendingGovernance`?

steffenix replied:
Copied from Registry, changing governance is 2 phases.

poolpitako replied:
this is for adding a new version of the strategy should be asserting governance not pending.

steffenix replied:
got it!

poolpitako commented on contracts/StrategyRegistry.vy:
```diff
@@ -0,0 +1,70 @@
+    assert msg.sender == self.pendingGovernance  # dev: unauthorized
+    strategyName: String[64] = name 
+    if name == "":
+        strategyName = Strategy(strategy).name()
```
tbh I would revert.

steffenix replied:
ok

poolpitako commented on contracts/StrategyRegistry.vy:
```diff
@@ -0,0 +1,70 @@
+	    self.strategiesReleases[strategyHash] = StrategyInfo({
+		    name: strategyName,
+		    numReleases: 0,
+		    versions: [ZERO_ADDRESS, ZERO_ADDRESS, ZERO_ADDRESS, ZERO_ADDRESS, ZERO_ADDRESS, ZERO_ADDRESS, ZERO_ADDRESS, ZERO_ADDRESS, ZERO_ADDRESS, ZERO_ADDRESS, ZERO_ADDRESS, ZERO_ADDRESS, ZERO_ADDRESS, ZERO_ADDRESS, ZERO_ADDRESS, ZERO_ADDRESS, ZERO_ADDRESS, ZERO_ADDRESS, ZERO_ADDRESS, ZERO_ADDRESS]
```
this is how the registry works? can we have a map inside the map with key = `numReleases`?

steffenix replied:
the issue is we can't have a hashmap within a structure but @fubuloubu suggested an other structure.

poolpitako commented on contracts/StrategyRegistry.vy:
```diff
@@ -0,0 +1,70 @@
+    numReleases: uint256 = self.strategiesReleases[strategyHash].numReleases
+    self.strategiesReleases[strategyHash].numReleases = numReleases + 1
+    self.strategiesReleases[strategyHash].versions[numReleases] = strategy
+
```
Not sure how we are going to handle the cloning inside here since the parameters will be different.
Have you thought of that issue?

steffenix replied:
I don't think that is necessary. Find the strategy latest version and then use the clone function from the strategy

poolpitako replied:
well, if we are doing a registry I would like the registry to have the cloning logic.
Perhaps a factory contract? I am trying to trim as much fat as possible from strategies.

salazarguille replied:
I think the complicated part is that each strategy has its own params for the cloning logic. Isn't it?

poolpitako replied:
yep, that's why I was thinking that perhaps the registry can map to a factory.
Something like: `registry.getFactory(likeStrategy)`

salazarguille replied:
We also might decode the params and use a `bytes data` as an input param in the `cloneStrategy` function. For example, using a struct we might

```MyCustomStruct memory mcs = abi.decode(data, (MyCustomStruct));````

salazarguille replied:
Just an idea: 
In case a strategy is clonable:
- We call the factory.cloneStrategy(...)
- Factory call the stategy.clone(data) function.
- The strategy should decode the data bytes.
- Clone the strategy.
- Initialize the cloned strategy with the data.
- Finally, the factory should register the new cloned strategy

steffenix replied:
So here I want to have two different registry, one that takes care of what's the most up to date version of a strategy (this contract) and an other one that can take care of keeping a list of all strategies (an other contract).

So the cloneStrategy should only be called by governance and then we can have a cloneStrategyExperiment function that can be done by anyone.

salazarguille commented on contracts/StrategyRegistry.vy:
```diff
@@ -0,0 +1,61 @@
+    @param governance The next governance address
+    """
+    assert msg.sender == self.governance  # dev: unauthorized
+    self.pendingGovernance = governance
```
I would emit an event when the `pendingGovernance` variable is changed, so we can track it offchain with thegraph.

steffenix replied:
So we have `log NewGovernance(msg.sender)`, what we have been doing with vaults is only emit when governance is accepted. Do you think `pendingGovernance` would also be a good thing to have  on the vault?

It seems to me it is.

salazarguille replied:
Yes, I think it would be good to have.

steffenix replied:
https://github.com/yearn/yearn-vaults/pull/412

salazarguille commented on contracts/StrategyRegistry.vy:
```diff
@@ -0,0 +1,61 @@
+        strategyName = Strategy(strategy).name()
+        assert strategyName != ""
+    apiVersion: String[8] = Strategy(strategy).apiVersion()
+    self.strategyVersions[StrategyInfo({name: strategyName, apiVersion: apiVersion})] = strategy
```
Should we emit an event here with strategy name, address, and apiVersion?

salazarguille commented on contracts/StrategyRegistry.vy:
```diff
@@ -0,0 +1,61 @@
+  name: String[64]
+  apiVersion: String[8]
+
+strategyVersions: public(HashMap[StrategyInfo, address])
```
AFAIK, to query this mapping we will always need the name and apiVersion to get the strategy address. Right?

steffenix replied:
Yes

steffenix commented:
Since last reviews:

- Added clone method to the StrategyVersionRegistry that slims down the BaseStrategy.
- Prevent clones of clones
- Removed initialized without params (can't find a way to do `abi.encoded(msg.sender,...)`)

fubuloubu commented on contracts/StrategyVersionRegistry.vy:
```diff
@@ -0,0 +1,85 @@
+interface Strategy:
+    def name() -> String[64]: view
+    def apiVersion() -> String[8]: view
+    def initialize(params: Bytes[128]): nonpayable
```
You can make this larger for future use (this is a dynamic variable, so this means "up to 256 bytes")
```suggestion
    def initialize(params: Bytes[256]): nonpayable
```

fubuloubu commented on contracts/StrategyVersionRegistry.vy:
```diff
@@ -0,0 +1,85 @@
+        strategyName = Strategy(strategy).name()
+        assert strategyName != ""
+    apiVersion: String[8] = Strategy(strategy).apiVersion()
+    key: String[73] = concat(strategyName, SEPARATOR, apiVersion)
```
Might be better to abstract this to it's own internal function
```python
@pure
@internal
def _computeKey(name: String[64], version: String[8]) -> String[73]:
    return concat(name, "@", version)  # NOTE: No need for constant here
```

And then just use it via:
```suggestion
    key: String[73] = self._computeKey(strategyName, apiVersion)
```

fubuloubu commented on contracts/StrategyVersionRegistry.vy:
```diff
@@ -0,0 +1,91 @@
+    return self.strategyVersions[key]
+
+@external
+def clone(strategy: address, params: Bytes[128]):
```
To match above change
```suggestion
def clone(strategy: address, params: Bytes[256]):
```

jmonteer commented on contracts/StrategyVersionRegistry.vy:
```diff
@@ -0,0 +1,90 @@
+        assert strategyName != ""
+    apiVersion: String[8] = Strategy(strategy).apiVersion()
+    key: String[73] = self._computeKey(strategyName, apiVersion)
+    self.strategyVersions[key] = strategy
```
does it make sense to check if it is already set? 

thinking about issues that may come from overwriting a strategy@version with a different deployment

steffenix replied:
It's expected to be overwritten as new version of a strategy are released.

jmonteer replied:
oh sure, I misinterpreted apiVersion for strategyVersion when writing this

still, any issues that may come from releasing a new version for the same StrategyName@ApiVersion pair? 

like getting all the existing clones from events and having 2 different deployments for same pair?

steffenix replied:
Sorry didn't got it

jmonteer replied:
trying to think about potential issues that may arise from the following process
I release a strategy for yvDAI 0.4.2 
I add it to the registry StrategyA@0.4.2
Then I find a bug in that strategy and redeploy
I need to add it to the registry, the key will still be StrategyA@0.4.2 and it will override previous strategy

might this cause any issue ? 
cc @salazarguille

steffenix replied:
That's the point is you keep updating the strategy address.

jmonteer commented on contracts/StrategyVersionRegistry.vy:
```diff
@@ -0,0 +1,90 @@
+
+@external
+def clone(strategy: address, params: Bytes[256]):
+    assert Strategy(strategy).isOriginal()  # dev: not original
```
should we add this method in BaseStrategy?

variable isOriginal set at deployment time so strategists don't need to handle this

steffenix replied:
Instead of BaseStrategyInitializable ?
The strategy needs at least to be `Initializable` so I will keep it within the BaseStrategyInitializable.

jmonteer replied:
I missed that it is already in BaseStrategyInitializable 

all good

jmonteer commented on contracts/StrategyVersionRegistry.vy:
```diff
@@ -0,0 +1,90 @@
+ 
+    Strategy(newStrategy).initialize(params)
+
+    log Cloned(newStrategy)
```
as this is a registry, would it make sense to emit more info about original strat? 

original address
original name ? 
original version ?

steffenix replied:
cc @salazarguille

jmonteer commented on contracts/StrategyVersionRegistry.vy:
```diff
@@ -0,0 +1,90 @@
+    governance: address
+
+event StrategyRegistered:
+    strategy: address
```
would it make sense to add this as indexed arg?

steffenix replied:
cc @salazarguille

jmonteer commented on contracts/StrategyVersionRegistry.vy:
```diff
@@ -0,0 +1,90 @@
+    assert msg.sender == self.governance  # dev: unauthorized
+    assert Strategy(strategy).isOriginal() # dev: not original 
+    strategyName: String[64] = name 
+    if name == "":
```
Question: why checking if a newly initialized variable is empty? shouldn't it be in 100% of cases? 

in my ignorance, I would directly do:
```
strategyName = Strategy(strategy).name()
assert strategyName != ""
```

steffenix replied:
name is a parameter:
You can set it to override the strategy name otherwise it will default to the Strategy.name()

```
addNewRelease(strategy :address, name: String[64] = ""):
```

steffenix commented:
Should BaseStrategy and BaseStrategyInitializable be merged together?

jmonteer commented:
@poolpitako had an opinion on this.

steffenix commented:
@poolpitako ^
